File: examples/python/hello_smc.py
#!/usr/bin/env python3
"""examples/python/hello_smc.py — minimal Python example using the ctypes binding."""

import sys

# Allow running from the repository root without installing the package.
sys.path.insert(0, str(__file__).rsplit("/", 2)[1])

import smc


def main() -> None:
    print("2 + 3 * 4 =", smc.eval("2 + 3 * 4"))
    print("(1 + 2) ^ 3 =", smc.eval("(1 + 2) ^ 3"))
    print("7 // 2 =", smc.eval_int("7 / 2"))

    with smc.Context(level=2) as ctx:
        print("10 - 4 / 2 =", ctx.eval("10 - 4 / 2"))


if __name__ == "__main__":
    main()

File: examples/c/hello_smc.c
/* examples/c/hello_smc.c — minimal C example using the stub runtime */
#include "smc.h"
#include <stdio.h>

int main(void) {
    int rc = smc_init();
    if (rc != 0) {
        fprintf(stderr, "smc_init failed: %s\n", smc_error_string(rc));
        return 1;
    }

    double result;

    rc = smc_eval_double("2 + 3 * 4", &result);
    if (rc != 0) {
        fprintf(stderr, "eval failed: %s\n", smc_last_error()->message);
        smc_shutdown();
        return 1;
    }
    printf("2 + 3 * 4 = %f\n", result);

    rc = smc_eval_double("(1 + 2) ^ 3", &result);
    if (rc != 0) {
        fprintf(stderr, "eval failed: %s\n", smc_last_error()->message);
        smc_shutdown();
        return 1;
    }
    printf("(1 + 2) ^ 3 = %f\n", result);

    smc_context_t *ctx = smc_context_create(2);
    rc = smc_eval_double_with(ctx, "10 - 4 / 2", &result);
    if (rc != 0) {
        fprintf(stderr, "context eval failed: %s\n", smc_last_error()->message);
        smc_context_destroy(ctx);
        smc_shutdown();
        return 1;
    }
    printf("10 - 4 / 2 = %f\n", result);
    smc_context_destroy(ctx);

    smc_shutdown();
    return 0;
}
